File: database/database_test.go
// Copyright (c) 2026 Kiruba Sankar Swaminathan
//
// This source code is licensed under the MIT license found in the
// LICENSE file in the root of this source tree.

package database

import (
	"context"
	"encoding/binary"
	"hash/crc32"
	"io"
	"log/slog"
	"math"
	"os"
	"path/filepath"
	"testing"
	"time"

	"turnstone/protocol"
)

func walLogicalUsed(t *testing.T, f *os.File) int64 {
	t.Helper()
	stat, err := f.Stat()
	if err != nil {
		t.Fatal(err)
	}
	if stat.Size() < 16 {
		return stat.Size()
	}
	buf := make([]byte, 16)
	if _, err := f.ReadAt(buf, stat.Size()-16); err != nil {
		t.Fatal(err)
	}
	if binary.BigEndian.Uint32(buf[0:4]) != 0x54534631 {
		return stat.Size()
	}
	return int64(binary.BigEndian.Uint64(buf[4:12]))
}

func putKV(t *testing.T, db *Database, key, val string) {
	t.Helper()
	tx := db.NewTransaction(true)
	if err := tx.Put([]byte(key), []byte(val)); err != nil {
		tx.Discard()
		t.Fatalf("put %s: %v", key, err)
	}
	if err := tx.Commit(); err != nil {
		t.Fatalf("commit %s: %v", key, err)
	}
}

func TestDatabase_Recover_Basic(t *testing.T) {
	dir := t.TempDir()
	logger := slog.New(slog.NewTextHandler(io.Discard, nil))

	s1, err := Open(context.Background(), dir, logger, 0, "none", 90, 0)
	if err != nil {
		t.Fatalf("Failed to create initial database: %v", err)
	}

	keys := []string{"alpha", "beta", "gamma"}
	for _, k := range keys {
		putKV(t, s1, k, "val-"+k)
	}

	if err := s1.Close(); err != nil {
		t.Fatalf("Failed to close database 1: %v", err)
	}

	s2, err := Open(context.Background(), dir, logger, 0, "none", 90, 0)
	if err != nil {
		t.Fatalf("Failed to create recovered database: %v", err)
	}
	defer s2.Close()

	for _, k := range keys {
		val, err := s2.Get(k)
		if err != nil {
			t.Errorf("Failed to get key %s: %v", k, err)
		}
		expected := "val-" + k
		if string(val) != expected {
			t.Errorf("Key %s: expected %s, got %s", k, expected, string(val))
		}
	}
}

func TestDatabase_Recover_CRC_Corruption(t *testing.T) {
	os.Setenv("TS_TEST_LOG_TRUNCATE", "true")
	defer os.Unsetenv("TS_TEST_LOG_TRUNCATE")
	dir := t.TempDir()
	logger := slog.New(slog.NewTextHandler(io.Discard, nil))

	s1, err := Open(context.Background(), dir, logger, 0, "none", 90, 0)
	if err != nil {
		t.Fatal(err)
	}
	putKV(t, s1, "key1", "val1")
	putKV(t, s1, "key2", "val2")
	s1.Close()

	logPath := filepath.Join(dir, "wal", "seg-000001.wal")
	if _, err := os.Stat(logPath); err != nil {
		t.Fatalf("No wal segment found in %s: %v", dir, err)
	}

	f, err := os.OpenFile(logPath, os.O_RDWR, 0o644)
	if err != nil {
		t.Fatal(err)
	}
	used := walLogicalUsed(t, f)
	if used < 2 {
		t.Fatalf("expected WAL used bytes, got %d", used)
	}
	b := make([]byte, 1)
	if _, err := f.ReadAt(b, used-1); err != nil {
		t.Fatal(err)
	}
	b[0] ^= 0xFF
	if _, err := f.WriteAt(b, used-1); err != nil {
		t.Fatal(err)
	}
	f.Close()

	s2, err := Open(context.Background(), dir, logger, 0, "none", 90, 0)
	if err != nil {
		t.Fatalf("Failed to recover database: %v", err)
	}
	defer s2.Close()

	val, err := s2.Get("key1")
	if err != nil {
		t.Errorf("Expected key1 to survive corruption, got error: %v", err)
	}
	if string(val) != "val1" {
		t.Errorf("Expected val1, got %s", val)
	}

	_, err = s2.Get("key2")
	if err != protocol.ErrKeyNotFound {
		t.Errorf("Expected key2 to be dropped due to CRC failure, got: %v", err)
	}
}

func TestDatabase_Recover_PartialWrite(t *testing.T) {
	os.Setenv("TS_TEST_LOG_TRUNCATE", "true")
	defer os.Unsetenv("TS_TEST_LOG_TRUNCATE")
	dir := t.TempDir()
	logger := slog.New(slog.NewTextHandler(io.Discard, nil))

	s1, err := Open(context.Background(), dir, logger, 0, "none", 90, 0)
	if err != nil {
		t.Fatal(err)
	}
	putKV(t, s1, "key1", "val1")
	s1.Close()

	logPath := filepath.Join(dir, "wal", "seg-000001.wal")
	f, err := os.OpenFile(logPath, os.O_RDWR, 0o644)
	if err != nil {
		t.Fatal(err)
	}
	used := walLogicalUsed(t, f)
	if _, err := f.WriteAt([]byte{0xFF, 0xFF, 0xFF, 0xFF}, used); err != nil {
		t.Fatal(err)
	}
	buf := make([]byte, 16)
	binary.BigEndian.PutUint32(buf[0:4], 0x54534631)
	binary.BigEndian.PutUint64(buf[4:12], uint64(used+4))
	binary.BigEndian.PutUint32(buf[12:16], crc32.Checksum(buf[:12], crc32.MakeTable(crc32.Castagnoli)))
	stat, err := f.Stat()
	if err != nil {
		t.Fatal(err)
	}
	if _, err := f.WriteAt(buf, stat.Size()-16); err != nil {
		t.Fatal(err)
	}
	f.Close()

	s2, err := Open(context.Background(), dir, logger, 0, "none", 90, 0)
	if err != nil {
		t.Fatalf("Recovery failed on partial write: %v", err)
	}
	defer s2.Close()

	if _, err := s2.Get("key1"); err != nil {
		t.Error("key1 lost during partial write recovery")
	}
}

func TestDatabase_Replication_Quorum(t *testing.T) {
	dir := t.TempDir()
	logger := slog.New(slog.NewTextHandler(io.Discard, nil))

	s, err := Open(context.Background(), dir, logger, 1, "none", 90, 0)
	if err != nil {
		t.Fatal(err)
	}
	defer s.Close()

	s.RegisterReplica("replica-1", 0, "server")

	done := make(chan error)
	go func() {
		tx := s.NewTransaction(true)
		if err := tx.Put([]byte("k"), []byte("v")); err != nil {
			tx.Discard()
			done <- err
			return
		}
		if err := tx.Commit(); err != nil {
			done <- err
			return
		}
		done <- s.WaitForQuorum(s.LastLogOffset(), 0, nil)
	}()

	select {
	case err := <-done:
		t.Fatalf("Write returned before quorum was met. Err: %v", err)
	case <-time.After(100 * time.Millisecond):
	}

	s.UpdateReplicaOffset("replica-1", s.LastLogOffset())

	select {
	case err := <-done:
		if err != nil {
			t.Fatalf("Write failed: %v", err)
		}
	case <-time.After(1 * time.Second):
		t.Fatal("Write timed out after quorum met")
	}
}

func TestDatabase_CommitPuts(t *testing.T) {
	dir := t.TempDir()
	logger := slog.New(slog.NewTextHandler(io.Discard, nil))

	s, err := Open(context.Background(), dir, logger, 0, "none", 90, 0)
	if err != nil {
		t.Fatal(err)
	}
	defer s.Close()

	putKV(t, s, "k1", "v1")
	putKV(t, s, "k2", "v2")

	val, err := s.Get("k1")
	if err != nil {
		t.Fatal(err)
	}
	if string(val) != "v1" {
		t.Errorf("Want v1, got %s", val)
	}

	val2, err := s.Get("k2")
	if err != nil {
		t.Fatal(err)
	}
	if string(val2) != "v2" {
		t.Errorf("Want v2, got %s", val2)
	}
}

func TestStats_ConflictsAndStorage(t *testing.T) {
	dir := t.TempDir()
	logger := slog.New(slog.NewTextHandler(io.Discard, nil))
	s, err := Open(context.Background(), dir, logger, 0, "none", 90, 0)
	if err != nil {
		t.Fatalf("Open failed: %v", err)
	}
	defer s.Close()

	stats := s.Stats()
	if stats.Conflicts != 0 {
		t.Errorf("expected 0 conflicts, got %d", stats.Conflicts)
	}

	tx1 := s.DB.NewTransaction(true)
	tx1.Put([]byte("key"), []byte("val1"))

	tx2 := s.DB.NewTransaction(true)
	tx2.Put([]byte("key"), []byte("val2"))

	if err := tx1.Commit(); err != nil {
		t.Fatalf("tx1 commit failed: %v", err)
	}

	if err := tx2.Commit(); err == nil {
		t.Error("tx2 should have failed with conflict")
	}

	stats = s.Stats()
	if stats.Conflicts != 1 {
		t.Errorf("expected 1 conflict, got %d", stats.Conflicts)
	}

	putKV(t, s, "k1", "v1")
	putKV(t, s, "k2", "v2")

	stats = s.Stats()
	if stats.LogSize == 0 {
		t.Error("expected >0 bytes log logical size")
	}
	if stats.LogSize != stats.Offset {
		t.Errorf("before truncation log_bytes should equal offset, log=%d offset=%d", stats.LogSize, stats.Offset)
	}
	if stats.LogAllocated == 0 {
		t.Error("expected >0 allocated log bytes")
	}

	detail := s.StorageDetail()
	if len(detail.Segments) == 0 {
		t.Fatal("expected at least one wal segment")
	}
	if detail.WalLiveBytes == 0 {
		t.Fatal("expected wal live bytes after writes")
	}
	if detail.IndexAllocatedBytes <= detail.IndexLiveBytes {
		t.Fatalf("allocated buffers should exceed live payload, allocated=%d live=%d", detail.IndexAllocatedBytes, detail.IndexLiveBytes)
	}
	if detail.IndexArenaBytes < detail.IndexLiveBytes {
		t.Fatalf("bump arena should be at least live payload, arena=%d live=%d", detail.IndexArenaBytes, detail.IndexLiveBytes)
	}
	if detail.HashShards < 1 {
		t.Fatalf("expected nonempty hash shards, got %d", detail.HashShards)
	}
	hash := s.IndexHashMetrics()
	if hash.ShardsUsed != detail.HashShards || hash.ArenaBytes != detail.IndexArenaBytes || hash.AllocatedBytes != detail.IndexAllocatedBytes || hash.LiveBytes != detail.IndexLiveBytes {
		t.Fatalf("IndexHashMetrics mismatch: %+v detail shards=%d arena=%d allocated=%d live=%d", hash, detail.HashShards, detail.IndexArenaBytes, detail.IndexAllocatedBytes, detail.IndexLiveBytes)
	}
	if s.WalSegmentCount() != len(detail.Segments) {
		t.Fatalf("segment count %d != %d", s.WalSegmentCount(), len(detail.Segments))
	}
}

func TestDatabase_ReplicaLag(t *testing.T) {
	dir := t.TempDir()
	logger := slog.New(slog.NewTextHandler(io.Discard, nil))
	s, err := Open(context.Background(), dir, logger, 0, "none", 90, 0)
	if err != nil {
		t.Fatalf("Open failed: %v", err)
	}
	defer s.Close()

	putKV(t, s, "k1", "v1")
	head := s.LastLogOffset()

	s.RegisterReplica("r1", head, "server")
	stats := s.Stats()
	if stats.ReplicaLag != 0 {
		t.Errorf("expected 0 lag, got %d", stats.ReplicaLag)
	}

	putKV(t, s, "k1", "v1")
	newHead := s.LastLogOffset()

	stats = s.Stats()
	expectedLag := newHead - head
	if stats.ReplicaLag != expectedLag {
		t.Errorf("expected lag %d, got %d", expectedLag, stats.ReplicaLag)
	}
	if len(stats.Replicas) != 1 {
		t.Fatalf("expected 1 replica in stats, got %d", len(stats.Replicas))
	}
	if stats.Replicas[0].ID != "r1" || stats.Replicas[0].Lag != expectedLag {
		t.Fatalf("unexpected replica info: %+v", stats.Replicas[0])
	}
	if stats.ServerReplicas != 1 {
		t.Fatalf("expected 1 server replica, got %d", stats.ServerReplicas)
	}
}

func TestDatabase_ReplicaLag_ReportsSlowestReplica(t *testing.T) {
	dir := t.TempDir()
	logger := slog.New(slog.NewTextHandler(io.Discard, nil))
	s, err := Open(context.Background(), dir, logger, 0, "none", 90, 0)
	if err != nil {
		t.Fatal(err)
	}
	defer s.Close()

	putKV(t, s, "k1", "v1")
	head := s.LastLogOffset()

	s.RegisterReplica("caught-up", head, "server")
	s.RegisterReplica("slow", head/2, "server")

	putKV(t, s, "k2", "v2")
	newHead := s.LastLogOffset()
	wantSlowLag := newHead - head/2

	stats := s.Stats()
	if stats.ReplicaLag != wantSlowLag {
		t.Fatalf("replica_lag=%d want slowest lag %d", stats.ReplicaLag, wantSlowLag)
	}
	if len(stats.Replicas) != 2 {
		t.Fatalf("expected 2 replicas, got %d", len(stats.Replicas))
	}
	if stats.ServerReplicas != 2 {
		t.Fatalf("expected 2 server replicas, got %d", stats.ServerReplicas)
	}
}

func TestDatabase_ReplicaLag_IgnoresNonServerRoles(t *testing.T) {
	dir := t.TempDir()
	logger := slog.New(slog.NewTextHandler(io.Discard, nil))
	s, err := Open(context.Background(), dir, logger, 0, "none", 90, 0)
	if err != nil {
		t.Fatal(err)
	}
	defer s.Close()

	putKV(t, s, "k1", "v1")
	head := s.LastLogOffset()
	s.RegisterReplica("backup", head/2, ReplicaRoleBackup)
	s.RegisterReplica("admin", head/3, ReplicaRoleAdmin)

	putKV(t, s, "k2", "v2")
	stats := s.Stats()
	if stats.ReplicaLag != 0 {
		t.Fatalf("backup/admin should not set replica_lag, got %d", stats.ReplicaLag)
	}
	if stats.ServerReplicas != 0 {
		t.Fatalf("expected 0 server replicas, got %d", stats.ServerReplicas)
	}

	s.RegisterReplica("server", head, ReplicaRoleServer)
	stats = s.Stats()
	newHead := s.LastLogOffset()
	want := uint64(newHead) - uint64(head)
	if stats.ReplicaLag != want {
		t.Fatalf("replica_lag=%d want server lag %d", stats.ReplicaLag, want)
	}
	if stats.ServerReplicas != 1 {
		t.Fatalf("expected 1 server replica, got %d", stats.ServerReplicas)
	}
}

func TestIsValidReplicationCursor_HeadAndZero(t *testing.T) {
	dir := t.TempDir()
	logger := slog.New(slog.NewTextHandler(io.Discard, nil))
	s, err := Open(context.Background(), dir, logger, 0, "none", 90, 0)
	if err != nil {
		t.Fatal(err)
	}
	defer s.Close()

	putKV(t, s, "k", "v")
	head := uint64(s.LastLogOffset())

	if !s.IsValidReplicationCursor(0) {
		t.Fatal("offset 0 should be valid for full sync")
	}
	if !s.IsValidReplicationCursor(head) {
		t.Fatal("head offset should be valid")
	}
	if s.IsValidReplicationCursor(head + 1) {
		t.Fatal("offset beyond head should be invalid")
	}
}

func TestIsValidReplicationCursor_RejectsMidFrame(t *testing.T) {
	dir := t.TempDir()
	logger := slog.New(slog.NewTextHandler(io.Discard, nil))
	s, err := Open(context.Background(), dir, logger, 0, "none", 90, 0)
	if err != nil {
		t.Fatal(err)
	}
	defer s.Close()

	putKV(t, s, "k", "v")
	head := s.LastLogOffset()
	if head <= 1 {
		t.Fatalf("expected head > 1, got %d", head)
	}
	if s.IsValidReplicationCursor(1) {
		t.Fatal("mid-frame offset should be rejected")
	}
	if !s.DB.IsValidFrameOffset(0) {
		t.Fatal("frame boundary at 0 expected")
	}
}

func TestMinReplicaOffset_ExcludesBackupRole(t *testing.T) {
	dir := t.TempDir()
	logger := slog.New(slog.NewTextHandler(io.Discard, nil))
	s, err := Open(context.Background(), dir, logger, 0, "none", 90, 0)
	if err != nil {
		t.Fatal(err)
	}
	defer s.Close()

	putKV(t, s, "k", "v")
	head := uint64(s.LastLogOffset())

	s.RegisterReplica("backup", head/2, ReplicaRoleBackup)
	if got := s.MinReplicaOffset(); got != math.MaxUint64 {
		t.Fatalf("backup role should not pin retention, got %d", got)
	}

	s.RegisterReplica("server", head/3, ReplicaRoleServer)
	if got := s.MinReplicaOffset(); got != head/3 {
		t.Fatalf("server role min offset = %d, want %d", got, head/3)
	}
}

func TestWaitForQuorum_IgnoresNonServerRoles(t *testing.T) {
	dir := t.TempDir()
	logger := slog.New(slog.NewTextHandler(io.Discard, nil))
	s, err := Open(context.Background(), dir, logger, 1, "none", 90, 0)
	if err != nil {
		t.Fatal(err)
	}
	defer s.Close()

	putKV(t, s, "k", "v")
	target := uint64(s.LastLogOffset())

	s.RegisterReplica("admin-only", target, ReplicaRoleAdmin)
	s.RegisterReplica("backup-only", target, ReplicaRoleBackup)

	done := make(chan error, 1)
	go func() {
		done <- s.WaitForQuorum(target, 200*time.Millisecond, nil)
	}()

	select {
	case err := <-done:
		if err == nil {
			t.Fatal("expected quorum timeout when only non-server roles acked")
		}
	case <-time.After(1 * time.Second):
		t.Fatal("WaitForQuorum did not return")
	}

	s.RegisterReplica("server", target, ReplicaRoleServer)
	if err := s.WaitForQuorum(target, time.Second, nil); err != nil {
		t.Fatalf("server role should satisfy quorum: %v", err)
	}
}

func TestDatabase_BasicInit(t *testing.T) {
	dir := t.TempDir()
	logger := slog.New(slog.NewTextHandler(io.Discard, nil))

	s, err := Open(context.Background(), dir, logger, 0, "none", 90, 0)
	if err != nil {
		t.Fatalf("Failed to create database: %v", err)
	}
	defer s.Close()

	putKV(t, s, "cache_test", "value")

	val, err := s.Get("cache_test")
	if err != nil {
		t.Fatalf("Read failed: %v", err)
	}
	if string(val) != "value" {
		t.Errorf("Unexpected value: %s", val)
	}
}

File: rust/crates/turnstone-database/tests/database_test.rs
// Copyright (c) 2026 Kiruba Sankar Swaminathan
//
// This source code is licensed under the MIT license found in the
// LICENSE file in the root of this source tree.

use std::sync::Arc;
use std::time::Duration;

use turnstone_database::{
    open, DatabaseError, OpenOptions, REPLICA_ROLE_ADMIN, REPLICA_ROLE_BACKUP, REPLICA_ROLE_SERVER,
};

fn put_kv(db: &turnstone_database::Database, key: &str, val: &str) {
    let mut tx = db.new_transaction(true);
    tx.put(key.as_bytes(), val.as_bytes()).unwrap();
    tx.commit().unwrap();
}

#[test]
fn database_recover_basic() {
    let dir = tempfile::tempdir().unwrap();
    {
        let s1 = open(&dir, OpenOptions::default()).unwrap();
        for k in ["alpha", "beta", "gamma"] {
            put_kv(&s1, k, &format!("val-{k}"));
        }
        s1.close().unwrap();
    }
    let s2 = open(&dir, OpenOptions::default()).unwrap();
    for k in ["alpha", "beta", "gamma"] {
        let val = s2.get(k).unwrap();
        assert_eq!(String::from_utf8_lossy(&val), format!("val-{k}"));
    }
    s2.close().unwrap();
}

#[test]
fn database_reset_wipes_keys() {
    let dir = tempfile::tempdir().unwrap();
    let s = open(&dir, OpenOptions::default()).unwrap();
    put_kv(&s, "keep-me", "until-reset");
    s.reset().unwrap();
    assert!(s.get("keep-me").is_err());
    s.close().unwrap();
}

#[test]
fn database_replication_quorum() {
    let dir = tempfile::tempdir().unwrap();
    let s = open(
        &dir,
        OpenOptions {
            min_replicas: 1,
            ..Default::default()
        },
    )
    .unwrap();
    s.register_replica("replica-1", 0, REPLICA_ROLE_SERVER);

    let s2 = Arc::clone(&s);
    let handle = std::thread::spawn(move || {
        let mut tx = s2.new_transaction(true);
        tx.put(b"k", b"v").unwrap();
        tx.commit().unwrap();
        s2.wait_for_quorum(s2.last_log_offset(), Duration::ZERO, None)
    });

    std::thread::sleep(Duration::from_millis(100));
    assert!(handle.is_finished() == false);

    s.update_replica_offset("replica-1", s.last_log_offset());
    handle.join().unwrap().unwrap();
}

#[test]
fn database_commit_puts() {
    let dir = tempfile::tempdir().unwrap();
    let s = open(&dir, OpenOptions::default()).unwrap();
    put_kv(&s, "k1", "v1");
    put_kv(&s, "k2", "v2");
    assert_eq!(String::from_utf8_lossy(&s.get("k1").unwrap()), "v1");
    assert_eq!(String::from_utf8_lossy(&s.get("k2").unwrap()), "v2");
}

#[test]
fn stats_conflicts_and_storage() {
    let dir = tempfile::tempdir().unwrap();
    let s = open(&dir, OpenOptions::default()).unwrap();
    assert_eq!(s.stats().conflicts, 0);

    let mut tx1 = s.new_transaction(true);
    tx1.put(b"key", b"val1").unwrap();
    let mut tx2 = s.new_transaction(true);
    let _ = tx2.put(b"key", b"val2");
    tx1.commit().unwrap();
    assert!(tx2.commit().is_err());
    assert_eq!(s.stats().conflicts, 1);

    put_kv(&s, "k1", "v1");
    put_kv(&s, "k2", "v2");
    let stats = s.stats();
    assert!(stats.log_size > 0);
    assert_eq!(stats.log_size, stats.offset);
    assert!(stats.log_allocated > 0);

    let detail = s.storage_detail();
    assert!(!detail.segments.is_empty());
    assert!(detail.wal_live_bytes > 0);
    assert!(detail.index_allocated_bytes > detail.index_live_bytes);
    assert!(detail.index_arena_bytes >= detail.index_live_bytes);
    assert!(detail.hash_shards >= 1);
    let hash = s.index_hash_metrics();
    assert_eq!(hash.shards_used, detail.hash_shards);
    assert_eq!(s.wal_segment_count(), detail.segments.len() as i32);
}

#[test]
fn database_replica_lag() {
    let dir = tempfile::tempdir().unwrap();
    let s = open(&dir, OpenOptions::default()).unwrap();
    put_kv(&s, "k1", "v1");
    let head = s.last_log_offset();
    s.register_replica("r1", head, REPLICA_ROLE_SERVER);
    assert_eq!(s.stats().replica_lag, 0);
    put_kv(&s, "k1", "v1");
    let new_head = s.last_log_offset();
    let stats = s.stats();
    assert_eq!(stats.replica_lag, new_head - head);
    assert_eq!(stats.replicas.len(), 1);
    assert_eq!(stats.server_replicas, 1);
}

#[test]
fn is_valid_replication_cursor_head_and_zero() {
    let dir = tempfile::tempdir().unwrap();
    let s = open(&dir, OpenOptions::default()).unwrap();
    put_kv(&s, "k", "v");
    let head = s.last_log_offset();
    assert!(s.is_valid_replication_cursor(0));
    assert!(s.is_valid_replication_cursor(head));
    assert!(!s.is_valid_replication_cursor(head + 1));
}

#[test]
fn is_valid_replication_cursor_rejects_mid_frame() {
    let dir = tempfile::tempdir().unwrap();
    let s = open(&dir, OpenOptions::default()).unwrap();
    put_kv(&s, "k", "v");
    let head = s.last_log_offset();
    assert!(head > 1);
    assert!(!s.is_valid_replication_cursor(1));
    assert!(s.engine().is_valid_frame_offset(0));
}

#[test]
fn min_replica_offset_connected_backup_pins() {
    let dir = tempfile::tempdir().unwrap();
    let s = open(&dir, OpenOptions::default()).unwrap();
    put_kv(&s, "k", "v");
    let head = s.last_log_offset();
    s.register_replica("backup", head / 2, REPLICA_ROLE_BACKUP);
    assert_eq!(s.min_replica_offset(), head / 2);
    s.unregister_replica("backup");
    assert_eq!(s.min_replica_offset(), u64::MAX);
    s.register_replica("server", head / 3, REPLICA_ROLE_SERVER);
    assert_eq!(s.min_replica_offset(), head / 3);
}

#[test]
fn wait_for_quorum_ignores_non_server_roles() {
    let dir = tempfile::tempdir().unwrap();
    let s = open(
        &dir,
        OpenOptions {
            min_replicas: 1,
            ..Default::default()
        },
    )
    .unwrap();
    put_kv(&s, "k", "v");
    let target = s.last_log_offset();
    s.register_replica("admin-only", target, REPLICA_ROLE_ADMIN);
    s.register_replica("backup-only", target, REPLICA_ROLE_BACKUP);

    let s2 = Arc::clone(&s);
    let handle =
        std::thread::spawn(move || s2.wait_for_quorum(target, Duration::from_millis(200), None));
    assert!(matches!(
        handle.join().unwrap(),
        Err(DatabaseError::Other(_))
    ));

    s.register_replica("server", target, REPLICA_ROLE_SERVER);
    s.wait_for_quorum(target, Duration::from_secs(1), None)
        .unwrap();
}

#[test]
fn register_replica_reregister_preserves_connected() {
    let dir = tempfile::tempdir().unwrap();
    let s = open(
        &dir,
        OpenOptions {
            min_replicas: 1,
            ..Default::default()
        },
    )
    .unwrap();
    put_kv(&s, "k", "v");
    let head = s.last_log_offset();
    let (_, gen1) = s.register_replica_hello("r1", head, REPLICA_ROLE_SERVER);
    let (_, gen2) = s.register_replica_hello("r1", head, REPLICA_ROLE_SERVER);
    assert_ne!(gen1, gen2);
    s.unregister_replica_gen("r1", gen1);
    s.wait_for_quorum(head, Duration::from_secs(1), None)
        .unwrap();
}
